### PR TITLE
feat (api): receive hook from repository manager

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: engine/api/workflow_hook.go [hunk 1]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
+			vcsServer := repositoriesmanager.GetProjectVCSServer(p, node.Context.Application.VCSServer)
+			if vcsServer != nil {
+				client, errclient := repositoriesmanager.AuthorizedClient(api.mustDB(), api.Cache, vcsServer)
+				log.Warning("%v", client)
+				if errclient != nil {
+					return sdk.WrapError(errclient, "getWorkflowHookModelsHandler> Cannot get vcs client")
+				}
+				webHookInfo, errWH := repositoriesmanager.GetWebhooksInfos(client)
+				if errWH != nil {
+					return sdk.WrapError(errclient, "getWorkflowHookModelsHandler> Cannot get vcs web hook info")
```
Comment (bnjjj) [inline engine/api/workflow_hook.go]: errWH

File: engine/api/workflow_hook.go [hunk 2]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
 			return sdk.WrapError(err, "getWorkflowHookModelsHandler")
 		}
+
+		// Post processing  on repositoryWebHook
+		hasRepoManager := false
+		repoWebHookEnable := false
+		if node.Context.Application != nil && node.Context.Application.RepositoryFullname != "" {
+			hasRepoManager = true
+		}
+		if hasRepoManager {
```
Comment (fsamin) [inline engine/api/workflow_hook.go]: early return

File: engine/api/workflow_hook.go [hunk 3]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
+				webHookInfo, errWH := repositoriesmanager.GetWebhooksInfos(client)
+				if errWH != nil {
+					return sdk.WrapError(errWH, "getWorkflowHookModelsHandler> Cannot get vcs web hook info")
+				}
+				repoWebHookEnable = webHookInfo.WebhooksSupported && !webHookInfo.WebhooksDisabled
+
+			}
+		}
+
+		if !repoWebHookEnable {
```
Comment (fsamin) [inline engine/api/workflow_hook.go]: early return

File: engine/api/workflow_hook.go [hunk 4]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
+		repoWebHookEnable := false
+		if node.Context.Application != nil && node.Context.Application.RepositoryFullname != "" {
+			hasRepoManager = true
+		}
+		if hasRepoManager {
+			// Call VCS to know if repository allows webhook and get the configuration fields
+			vcsServer := repositoriesmanager.GetProjectVCSServer(p, node.Context.Application.VCSServer)
+			if vcsServer != nil {
+				client, errclient := repositoriesmanager.AuthorizedClient(api.mustDB(), api.Cache, vcsServer)
+				log.Warning("%v", client)
```
Comment (fsamin) [inline engine/api/workflow_hook.go]: remove

File: engine/api/workflow_hook.go [hunk 5]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
+				repoWebHookEnable = webHookInfo.WebhooksSupported && !webHookInfo.WebhooksDisabled
+
+			}
+		}
+
+		if !repoWebHookEnable {
+			var indexToDelete int
+			for i := range m {
+				if m[i].Name == workflow.RepositoryWebHookModel.Name {
+					indexToDelete = i
```
Comment (fsamin) [inline engine/api/workflow_hook.go]: break

File: engine/api/workflow_hook.go [hunk 6]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
+		key := vars["key"]
+		workflowName := vars["permWorkflowName"]
+		nodeID, errN := requestVarInt(r, "nodeID")
+		if errN != nil {
+			return sdk.WrapError(errN, "getWorkflowHookModelsHandler")
+		}
+
+		p, errP := project.Load(api.mustDB(), api.Cache, key, getUser(ctx))
+		if errP != nil {
+			return sdk.WrapError(errP, "getWorkflowHookModelsHandler")
```
Comment (fsamin) [inline engine/api/workflow_hook.go]: getWorkflowHookModelsHandler> Unable to load project

File: engine/api/workflow_hook.go [hunk 7]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
+		}
+
+		wf, errW := workflow.Load(api.mustDB(), api.Cache, key, workflowName, getUser(ctx))
+		if errW != nil {
+			return sdk.WrapError(errW, "getWorkflowHookModelsHandler")
```
Comment (fsamin) [inline engine/api/workflow_hook.go]: getWorkflowHookModelsHandler> Unable to load workflow

File: engine/api/workflow_hook.go [hunk 8]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
+		}
+
+		node := wf.GetNode(nodeID)
+		if node == nil {
+			return sdk.WrapError(sdk.ErrWorkflowNodeNotFound, "getWorkflowHookModelsHandler")
```
Comment (fsamin) [inline engine/api/workflow_hook.go]: getWorkflowHookModelsHandler> Unable to get node

File: engine/api/workflow_hook.go [hunk 9]
```diff
@@ -23,10 +26,68 @@ func (api *API) getWorkflowHooksHandler() Handler {
...
+		}
+
 		m, err := workflow.LoadHookModels(api.mustDB())
 		if err != nil {
 			return sdk.WrapError(err, "getWorkflowHookModelsHandler")
```
Comment (fsamin) [inline engine/api/workflow_hook.go]: getWorkflowHookModelsHandler> Unable to load hook models

File: engine/hooks/tasks.go [hunk 10]
```diff
@@ -293,6 +309,87 @@ func (s *Service) doScheduledTaskExecution(t *TaskExecution) (*sdk.WorkflowNodeR
...
+}
+
+func executeRepositoryWebHook(t *TaskExecution) (*sdk.WorkflowNodeRunHookEvent, error) {
+	// Prepare a struct to send to CDS API
+	h := sdk.WorkflowNodeRunHookEvent{
+		WorkflowNodeHookUUID: t.UUID,
+	}
+
+	payload := make(map[string]interface{})
+	if v, ok := t.WebHook.RequestHeader[GithubHeader]; ok && v[0] == "push" {
```
Comment (fsamin) [inline engine/hooks/tasks.go]: make each case a func and replace the if else if else if... by a switch case

File: engine/api/workflow_hook.go (hunk 2 above)
Comment (sguiheux) [reply engine/api/workflow_hook.go]: don't want to write multiple WriteJSON

File: engine/api/workflow_hook.go (hunk 3 above)
Comment (sguiheux) [reply engine/api/workflow_hook.go]: don't want to write multiple WriteJSON

File: engine/api/workflow_hook.go (hunk 9 above)
Comment (sguiheux) [reply engine/api/workflow_hook.go]: workflow.LoadHookModels already get this information

File: engine/api/workflow_hook.go (hunk 5 above)
Comment (sguiheux) [reply engine/api/workflow_hook.go]: done

File: engine/hooks/tasks.go [hunk 11]
```diff
@@ -315,14 +315,26 @@ func (s *Service) doWebHookExecution(t *TaskExecution) (*sdk.WorkflowNodeRunHook
 	return executeWebHook(t)
 }
 
+func getRepositoryHeader(whe *WebHookExecution) string {
+	if v, ok := whe.RequestHeader[GithubHeader]; ok && v[0] == "push" {
```
Comment (fsamin) [inline engine/hooks/tasks.go]: what about create event ? https://developer.github.com/v3/activity/events/types/#createevent

Comment (sguiheux) [reply engine/hooks/tasks.go]: it not represents a push event

Comment (sguiheux) [reply engine/hooks/tasks.go]: other event will arrive after

**Pull request merged**